File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        if(!root) {
            return false;
        }

        if(this.sameRoot(root,subRoot)) {
            return true;
        }

        return this.isSubtree(root.left , subRoot) || 
            this.isSubtree(root.right , subRoot);
    }
    sameRoot(root, subRoot) {
        if(!root && !subRoot) return true;
        if(!root || !subRoot) return false;

        return (root.val === subRoot.val) && 
            this.sameRoot(root.left , subRoot.left) &&
            this.sameRoot(root.right , subRoot.right);

    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameter = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        depth(root);
        return diameter;
    }

    int depth (TreeNode* root) {
        if(root == nullptr) {
            return 0;
        }

        int leftSide = depth(root -> left);
        int rightSide = depth(root -> right);

        diameter = std::max(diameter , leftSide + rightSide);

        return 1 + std::max(leftSide , rightSide);


    }
};
